plugin.rb

```diff
@@ -26,44 +29,83 @@ class Engine < ::Rails::Engine
       engine_name PLUGIN_NAME
       isolate_namespace DiscourseTranslator
     end
+
+    def self.current_service
+      "DiscourseTranslator::#{SiteSetting.translator}".constantize
+    end
   end
 
   class DiscourseTranslator::TranslatorController < ::ApplicationController
     before_action :ensure_logged_in
+    before_action :ensure_plugin_enabled
+    before_action :rate_limit
+    before_action :find_object
 
     def translate
-      raise PluginDisabled if !SiteSetting.translator_enabled
-
-      if !current_user.staff?
-        RateLimiter.new(current_user, "translate_post", SiteSetting.max_translations_per_minute, 1.minute).performed!
-      end
-
-      params.require(:post_id)
-      post = Post.find_by(id: params[:post_id])
-      raise Discourse::InvalidParameters.new(:post_id) if post.blank?
-      guardian.ensure_can_see!(post)
-
       begin
-        detected_lang, translation = "DiscourseTranslator::#{SiteSetting.translator}".constantize.translate(post)
+        detected_lang, translation = DiscourseTranslator.current_service.translate(@object)
         render json: { translation: translation, detected_lang: detected_lang }, status: 200
       rescue ::DiscourseTranslator::TranslatorError => e
         render_json_error e.message, status: 422
       end
     end
+
+    protected
+
+    def ensure_plugin_enabled
+      raise PluginDisabled if !SiteSetting.translator_enabled
+    end
+
+    def rate_limit
+      if !current_user.staff?
+        RateLimiter.new(current_user, "translate", SiteSetting.max_translations_per_minute, 1.minute).performed!
+      end
+    end
+
+    def find_object
+      raise Discourse::InvalidParameters.new unless params[:post_id] || params[:topic_id]
+      @object = params[:post_id] ? Post.find_by(id: params[:post_id]) : Topic.find_by(id: params[:topic_id])
```

Comment:
I'll prefer if we can just implement two different controller actions here. I'm not a big fan of having an API endpoint dealing with two different models because it leads to a bunch of conditional statements in the code down the line.